src/contrib/cluster/Akka.Cluster.Tools/Singleton/ClusterSingletonManager.cs

```diff
@@ -917,38 +1089,88 @@ private void InitializeFSM()
                 return null;
             });
 
-            When(ClusterSingletonState.Oldest, e =>
+            When(ClusterSingletonState.AcquiringLease, e =>
             {
-                if (e.FsmEvent is OldestChangedBuffer.OldestChanged oldestChanged && e.StateData is OldestData oldestData)
+                if (e.FsmEvent is AcquireLeaseResult alr)
                 {
-                    _oldestChangedReceived = true;
-                    Log.Info("Oldest observed OldestChanged: [{0} -> {1}]", _cluster.SelfAddress, oldestChanged.Oldest?.Address);
-                    switch (oldestChanged.Oldest)
+                    Log.Info("Acquire lease result {0}", alr.HoldingLease);
+                    if (alr.HoldingLease)
```

Comment:
So, removing a lot of this business logic from the cluster singleton implementation itself and moving it to the leasing system - ok, that makes sense.

Reply:
Oh I see - it's been moved to the state rather than the event-handling methods themselves.